### ISSUE TITLE
Completely deterministic network, but result is different from the past and different on different machines.

### ISSUE THREAD

**Issue by bugtig** [pytorch/pytorch#18412]
Hello, 

I have a model where I set
`   random.seed(config.seed)
    np.random.seed(config.seed)
    torch.manual_seed(config.seed)
    torch.cuda.manual_seed(config.seed)
    torch.cuda.manual_seed_all(config.seed)
`
and is therefore completely deterministic, and every time I run it it has the same input, gradients and results. 

However, now that deterministic gradients and results have now changed on my machine after a couple of weeks. For concreteness, the exact loss I should get at 100 iterations is 19.03 but the new result is 19.6. At first I thought it was due to libraries being installed and changing the original libraries, so I reinstalled everything, reinstalled linux, even. But every time, I get 19.6. 

So I tried running the same code on AWS, and the result is 19.03. After seeing this, I saw that even when running the same installation procedure, the AWS setting ends up with different versions of mkl, so now I copied the conda env from AWS and reproduced it on my machine, the result is still 19.6. I even tried a _different_ AWS instance, and the result is the same 19.03.

So finally I ran iterations side by side on AWS and my machine, and the reason for the difference is that they have slightly different gradients (order of e-5), and this quickly accumulates to the diverging results. 

To summarize: I kept getting 19.03 loss on my machine with a completely deterministic network, after a week I kept getting 19.6, tried a different machine on AWS, kept getting the original 19.03, completely wiped my machine and tried again, keep getting the same 19.6.

I'm having a hard time understanding what the issue might be. Is it some sort of hardware failure in my machine's GPU?

**Comment by SsnL** [comment]
https://pytorch.org/docs/stable/notes/randomness.html

**Comment by bugtig** [comment]
> https://pytorch.org/docs/stable/notes/randomness.html

Oh I should also note that I've tried `torch.backends.cudnn.deterministic = True
torch.backends.cudnn.benchmark = False` as well but they didn't solve the issue.

**Comment by SsnL** [comment]
Yeah. I'm more referring to this part:
> Completely reproducible results are not guaranteed across PyTorch releases, individual commits or different platforms. Furthermore, results need not be reproducible between CPU and GPU executions, even when using identical seeds.

**Comment by bugtig** [comment]
> Yeah. I'm more referring to this part:
> 
> > Completely reproducible results are not guaranteed across PyTorch releases, individual commits or different platforms. Furthermore, results need not be reproducible between CPU and GPU executions, even when using identical seeds.

Ah okay. Difference between different devices makes sense.. but do you have any sense why the results on my machine would have changed? (The same machine is producing different results after a while)

**Comment by SsnL** [comment]
Could be this
> There are some PyTorch functions that use CUDA functions that can be a source of non-determinism. One class of such CUDA functions are atomic operations, in particular atomicAdd, where the order of parallel additions to the same value is undetermined and, for floating-point variables, a source of variance in the result. PyTorch functions that use atomicAdd in the forward include torch.Tensor.index_add_(), torch.Tensor.scatter_add_(), torch.bincount().

> A number of operations have backwards that use atomicAdd, in particular torch.nn.functional.embedding_bag(), torch.nn.functional.ctc_loss() and many forms of pooling, padding, and sampling. There currently is no simple way of avoiding non-determinism in these functions.

**Comment by bugtig** [comment]
Where i'm confused is that every time I run the same code on any given machine, the results are completely deterministic (same gradient, same loss, same output), so it's not really non-deterministic in that sense, but it IS non-deterministic, to abuse the term, when comparing across devices/platforms. 

The real issue is that it's also non-deterministic across _time_ on my own machine, because I'm getting a different set of deterministic results than from before on the _same_ machine. But I guess there isn't much I can do about this right now...

**Comment by fmassa** [comment]
This is expected, some of our kernels are not deterministic (specially during backward).
Might be good to refer to https://github.com/pytorch/pytorch/issues/15359

I'm closing this issue in order to keep the discussion centralized.

**Comment by jhagege** [comment]
@bugtig having the exact same (weird) issue. 
Results used to be deterministic across platforms (running only on CPU) but lately they are only consistent on the same platform (same CPU) but not across different machines. 
Did you find any solution ? 
Thanks !

**Comment by jhagege** [comment]
+1 ?

**Comment by SsnL** [comment]
@jhagege yes, reproducible result is only guaranteed on same platform + same version.

**Comment by fmassa** [comment]
Also note that using multiple threads can also contribute to non-reproducibility
